R/model_ivpin.R

```diff
@@ -0,0 +1,905 @@
+  # (OI), the sample length (samplength) and the volume bucket size (vbs).
+  # Assume for now that samplength is 50. The formula for the first VPIN
+  # observation is sum(OI, i=1 to 50)/(50*vbs).
+  # The formula for the 5th VPIN observation is sum(OI, i=5 to 54)/(50*vbs).
+
+  # Note that sum(OI, i=5 to 54) = sum(OI, i=1 to 54) - sum(OI, i=1 to 4)
+  # The first one is the cumulative sum at 54 and the second one is cumulative
+  # sum at 4. So sum(OI, i=5 to 54) = cumsum[54]-cum[4].
+  # The general formula is sum(OI, i=n to 50+n) = cumsum[50+n]-cum[n].
+  #
+  # So we can calculate VPIN for the bucket 50, we shift the cumsum by 1
+  # position. We calculate first the cumulative sum for OI and then use the
+  # formula to find the value of VPIN
+
+  # --------------------------------------------------------------------------
+  # VII.1 CALCULATING VPIN VECTOR
+  # --------------------------------------------------------------------------
+
+  # Calculate the cumulative sum of OI: cumoi
+
+  if (nrow(bucketdata) < samplength) {
+
+    estimatevpin@success <- FALSE
+
+    estimatevpin@errorMessage <- vpin_err$largesamplength
+
+    ux$show(c= verbose, m = vpin_ms$aborted)
+
+    ux$show(ux$line())
+
+    ux$stopnow(m = vpin_err$largesamplength, s = vpin_err$fn)
+
+  }
+
+
+  # --------------------------------------------------------------------------
+  # VII.2 CORRECTING THE DURATION VECTOR
+  # --------------------------------------------------------------------------
+
+  corduration <- function(etime, stime, duration) {
+
+    days <- round(as.numeric(difftime(etime, stime, units = "days")))
+
+    duration <- as.numeric(duration)
+
+    if (days > 0) {
+      dseconds <- (days - 1) * 3600 * 24 + (24 - tradinghours) * 3600
+      return(duration - dseconds)
+    }
+    return(duration)
+  }
+
+  if (tradinghours > 0 & tradinghours < 24) {
+    bucketdata$duration <- apply(bucketdata, 1, function(x)
+      corduration(x[6], x[5], x[10]))
+  }
+  print(nrow(data))
+  # Normalize the duration to be centered around 1
+  mean_duration <- mean(bucketdata$duration)
+  bucketdata$duration <- bucketdata$duration / mean_duration
+  bucketdata$duration <- ifelse(bucketdata$duration == 0, 0.1, bucketdata$duration)
+                                
+  compute_log_likelihood <- function(params, t, Vb, Vs) {
+    alpha <- sigmoid(params[1])
+    delta <- sigmoid(params[2])
+    mu <- params[3] ^ 2
+    eps <- params[4] ^ 2
+    
+    e_1 <- log(alpha * delta) + Vb * log(eps) + Vs * log(eps + mu) - (2 * eps + mu) * t
+    e_2 <- log(alpha * (1 - delta)) + Vb * log(eps + mu) + Vs * log(eps) - (2 * eps + mu) * t
+    e_3 <- log(1 - alpha) + Vb * log(eps) + Vs * log(eps) - 2 * eps * t
+    e_max <- pmax(e_1, e_2, e_3)
+    
+    log_likelihood <- -sum(log(exp(e_1 - e_max) + exp(e_2 - e_max) + exp(e_3 - e_max)) + e_max)
+    return(log_likelihood)
+  }
+
+  # Compute iVPIN with a rolling window
+
+  start_index <- samplength + 1
+
+  "
```

Comment:
This part is well-commented but could be clarified a bit more so the reader knows why you calculate these variables. It helps to have consistent variable names: `total_arrival_rate` and `informed_arrival_rate`.
```R
# Easley et al. (2012b) derive the VPIN estimator based on two moment conditions from Poisson processes:
# E[|VB - VS|] ≈ alpha*mu and E[|VB + VS|] = 2*epsilon + alpha*mu.
# Ke et al. (2017) suggest that these conditions should be expressed as:
# E[|VB - VS||t; theta] ≈ alpha*mu*t and E[|VB + VS||t; theta] = (2*epsilon + alpha*mu)*t.
# These equations correspond to Equations (3) on page 364 of Ke et al. (2017) .
# 
# In our implementation:
# - The variable `total_arrival_rate` is calculated as  |VB + VS|/t. Its average (expected value) would, therefore, represent 2*epsilon + alpha*mu.
# - The variable `informed_arrival_rate` is calculated as |VB - VS|/t.  Its average (expected value) would, therefore, approximate alpha*mu.
# 
# This approximation allows us to estimate mu by dividing `informed_arrival_rate` by an estimated alpha.
# Once mu is estimated, we can determine epsilon using the equation for `total_arrival_rate`.

total_arrival_rate <- (bucketdata$agg.bvol + bucketdata$agg.svol) / bucketdata$duration
informed_arrival_rate <- abs(bucketdata$agg.bvol - bucketdata$agg.svol) / bucketdata$duration
```